### ISSUE TITLE
Migrating off pykg-config

### ISSUE THREAD

**Issue by dmacks** [fink/fink-distributions#918]
'pkg-config' is written in C and has dependencies including glib2 (which wants pkg-config or equivalent tool as a BuildDepends). For places where we don't want such a heavy toolchain (and to avoid a dep-cycle in glib2, something upstream is somewhat working on solving), we have 'pykg-gconfig' that is written in python (currently hardcoded to use /usr/bin/python) as an alternative. But now Apple is dropping that python interp. And python has a heavy dependency chain and and also has pkg-config or equivalent tool as a BuildDepends, so that's a different dep-cycle. So now we also have 'ppkg-config', written in perl and no deps except perl-core itself. So that should be a portable solution since we are guaranteed to have a perl interp at some known location.

I propose deprecating pykg-config, since it is currently little-used and has the most intractable dep-cycle, and switching to ppkg-config for places we don't want to use the regular pkg-config. Current pykg-config users:
 - [ ] arpack-ng
 - [x] glib2
 - [x] gphoto2
 - [x] libgphoto2.6
 - [x] libxml2
 - [ ] octave-3.8.2
 - [x] octave-3.6.4 (killed)
 - [x] pkgconfig
 - [ ] qrupdate
 - [x] xmkmf

**Comment by nieder** [comment]
Sounds reasonable. Make it so.

**Comment by dhomeier** [comment]
FWIW 1.3.0 seems to work with system python3 just out of the box, just needs a check for the executable in /usr/bin.
But if the perl alternative is easier to drop in, that's probably more maintainable. Otherwise I can open a PR to keep those other packages supported until they are migrated to ppkg-config.

**Comment by dmacks** [comment]
> FWIW 1.3.0 seems to work with system python3 just out of the box, just needs a check for the executable in /usr/bin. But if the perl alternative is easier to drop in, that's probably more maintainable.

If it were just that we needed to find the right executable-name on different platforms, that would be doable. But recent OSX dropped /usr/bin/python* altogether.

**Comment by dhomeier** [comment]
> If it were just that we needed to find the right executable-name on different platforms, that would be doable. But recent OSX dropped /usr/bin/python* altogether.

Uh, which ones are those? I tested this on Monterey 12.3 both for x86_64 (Rosetta) and arm64, and basically
```bash
        [ -x /usr/bin/python ] || pyversion=3
        /usr/bin/python$pyversion setup.py build_py --with-pc-path="$pc_path" build_scripts
```
just worked for pykg-config, as it fortunately already supports Python 3.
There are at least a dozen packages using /usr/bin/python in their build scripts, some of them unfortunately still on Python 2-only syntax, so it's already a pain to convert them (I have not been able to build python27 on arm64), but if Apple is dropping system-python entirely, that's a whole different can of worms... ˜\_(ツ)_/˜

**Comment by dmacks** [comment]
"Apple dropping system-python" was the word around other fink developers. My sandbox is pinned at 10.13:(

**Comment by dhomeier** [comment]
I am afraid I don't remember – had both installed before I started into any attempts to get Fink up.
I think I heard about Apple making no guarantees to continue providing a system Python in future OSs; so for this case migrating to ppkg-config may indeed be the safer option (is a system Perl guaranteed to remain available?)

**Comment by dmacks** [comment]
Fink will somehow guarantee a stable perl interp, given that fink itself is written in perl.

**Comment by dmacks** [comment]
Future glib will use the meson build system, which is written in python3, and therefore at least the python3X that meson uses will have to migrate from pkg-config to ppkg-config to avoid a dependency cycle (python310->pkgconfig->glib->meson->python310). Can probably handle it as part of PR #920

**Comment by eli-schwartz** [comment]
Linux distros have generally tended to migrate from Freedesktop pkg-config to https://github.com/pkgconf/pkgconf, another C implementation which does not, however, depend on glib, and builds with autotools.

pkgconf is generally considered to be better maintained than pkg-config, serves as a drop-in replacement, runs faster and supports more of the pkg-config spec anyway.

Personally, I advise against using some perl thing which, to my understanding, mainly exists for internal use by Strawberry Perl for the sake of more efficiently contaminating the default $PATH and screwing up MSVC by returning mingw libraries even when Strawberry Perl's bundled GCC isn't on the $PATH and isn't your compiler. :) On that note, if Meson detects that pkg-config is the perl implementation, it will log the following message, then return not-found:

```
found pkg-config '/path/to/pkg-config' but it is Strawberry Perl and thus broken. Ignoring...
```

So if you decide to use it, you will not be able to build any Meson software. :)

(No, we will not be enabling the use of perl pkg-config again, because it totally breaks things for any Windows user that installed perl.exe because e.g. the OpenSSL build instructions say to do so.)

**Comment by eli-schwartz** [comment]
Meson doesn't *technically* need pkg-config as a runtime dependency. It will only look for pkg-config if the software (such as glib) has any BDeps, which it then tries to find using pkg-config. Absence of pkg-config is not a fatal Meson error, but it does mean no dependencies will be found, and generally those dependencies are, themselves, fatal if not found.

In other words, glib has a BDeps on `['meson', 'pkg-config']`, and pkg-config has a [B]Deps on `['glib']`.

As I said, glib's BDeps on pkg-config can be satisfied by [pkgconf](https://github.com/pkgconf/pkgconf), and pkgconf has no BDeps.

**Comment by dhomeier** [comment]
Yes, probably should have been clearer that my one attempt at a real world usage of meson (building the new Scipy rc) failed with the above error when pointed to `%p/bin/ppkg-config`. It does accept `pykg-config` for the same build.
